query/src/main/java/com/connexta/replication/query/QueryService.java

```diff
@@ -0,0 +1,219 @@
+  private Site site;
+
+  private SiteQueue siteQueue;
+
+  /** The adapter to communicate with the given site. */
+  private NodeAdapter adapter;
+
+  @VisibleForTesting ScheduledFuture future;
+
+  /**
+   * Instantiates a new query service.
+   *
+   * @param site the configuration for the site to be queried by this query service
+   */
+  public QueryService(Site site, QueryServiceTools queryServiceTools) {
+    this(site, queryServiceTools, Executors.newSingleThreadScheduledExecutor());
+  }
+
+  public QueryService(
+      Site site, QueryServiceTools queryServiceTools, ScheduledExecutorService executor) {
+    this.queryServiceTools = queryServiceTools;
+    this.siteQueue = queryServiceTools.getQueueFor(site);
+    this.executor = executor;
+    update(site);
+  }
+
+  /**
+   * Updates the configuration for the site currently being queried by this query service.
+   *
+   * @param site the new site configuration
+   */
+  public void update(Site site) {
+    synchronized (this) {
+      this.site = site;
+      try {
+        this.adapter = queryServiceTools.getAdapterFor(site);
+      } catch (AdapterException e) {
+        LOGGER.info(
+            "Failed to create an adapter for {} stopping the query service for now."
+                + site.getUrl());
+        stop();
+      }
+    }
+  }
+
+  /** Starts this query service. */
+  public void start() {
+    if (future == null || future.isDone()) {
+      future =
+          executor.scheduleAtFixedRate(
+              this::query, 0, queryServiceTools.getPeriod(), TimeUnit.SECONDS);
+    }
+  }
+
+  /** Stops this query service. */
+  public void stop() {
+    executor.shutdownNow();
+  }
+
+  @VisibleForTesting
+  void query() {
+    if (!adapter.isAvailable()) {
+      LOGGER.debug("System at " + site.getUrl() + " is currently unavailable");
+      return;
+    }
+
+    Set<Filter> filters = queryServiceTools.activeFiltersFor(site).collect(Collectors.toSet());
+    for (Filter filter : filters) {
+      FilterIndex filterIndex = queryServiceTools.getOrCreateFilterIndex(filter);
+
+      // holds the modified time of each task put in the queue for this filter
+      // this will be used to grab the latest modified time after
+      Set<Instant> modifiedTimes = new HashSet<>();
+
+      try {
+        getNewMetadataForFilter(filter, filterIndex)
+            .map(metadatum -> metadataToTask(metadatum, filter.getPriority()))
+            .forEach(task -> queueTaskAndPutLastModifiedInSet(task, modifiedTimes));
+      } catch (AdapterException e) {
+        LOGGER.debug(
+            "Couldn't query site {}. Stopping queries for this site until next polling attempt.",
+            site.getName(),
+            e);
+        return;
+      } catch (QueueException qe) {
+        LOGGER.debug(
+            "Failed to put a task in queue for site {}. Stopping queuing of tasks for this site. Will resume at next polling attempt.",
+            site.getName(),
```

Comment:
❓ Should we still update the index for what we have do so far?

Reply:
Yes. good catch